### ISSUE TITLE
Registration of DefaultAccessibleProvider not done for C++ Builder

### ISSUE THREAD

**Issue by 7even11** [JAM-Software/Virtual-TreeView#951]
At the end of **VirtualTrees.Accessibility.pas** there are a few lines of code which register the **DefaultAccessibleProvider**, **DefaultAccessibleItemProvider** and **MultiColumnAccessibleProvider**. These are absent from the VirtualTrees.Accessibility.hpp so simply including the **VirtualTrees.Accessibility.hpp** in a C++ Builder project doesn't do anything and doesn't provide any accessibility. This should be included so that the VirtualTrees can provide the same accessibility support which it does in Delphi after including the **VirtualTrees.Accessibility** in the **uses** section.

The code in question is this one:

> initialization
  if DefaultAccessibleProvider = nil then
  begin
    DefaultAccessibleProvider := TVTDefaultAccessibleProvider.Create;
    GetAccessibilityFactory.RegisterAccessibleProvider(DefaultAccessibleProvider);
  end;
  if DefaultAccessibleItemProvider = nil then
  begin
    DefaultAccessibleItemProvider := TVTDefaultAccessibleItemProvider.Create;
    GetAccessibilityFactory.RegisterAccessibleProvider(DefaultAccessibleItemProvider);
  end;
  if MultiColumnAccessibleProvider = nil then
  begin
    MultiColumnAccessibleProvider := TVTMultiColumnAccessibleItemProvider.Create;
    GetAccessibilityFactory.RegisterAccessibleProvider(MultiColumnAccessibleProvider);
  end;
finalization
  GetAccessibilityFactory.UnRegisterAccessibleProvider(MultiColumnAccessibleProvider);
  MultiColumnAccessibleProvider := nil;
  GetAccessibilityFactory.UnRegisterAccessibleProvider(DefaultAccessibleItemProvider);
  DefaultAccessibleItemProvider := nil;
  GetAccessibilityFactory.UnRegisterAccessibleProvider(DefaultAccessibleProvider);
  DefaultAccessibleProvider := nil;

end.

**Comment by joachimmarder** [comment]
> At the end of VirtualTrees.Accessibility.pas there are a few lines of code which register the DefaultAccessibleProvider, DefaultAccessibleItemProvider and MultiColumnAccessibleProvider. These are absent from the VirtualTrees.Accessibility.hpp

This is to be expected. The .HPP-file is not a fully translated .PAS-file, this C++ header file is just the translated interface section of the .PAS-file. C++ Builder compiles the .PAS-files using the Delphi compiler and links it into the final binary.

So it would be interesting to find out if the code in question is actually executed. Can you put a breakpoint or `ShowMessage()` in  `TVTDefaultAccessibleItemProvider.Create()` and see it is executed?

**Comment by 7even11** [comment]
Thank you for looking into this.

Yes, for Delphi program which uses `TVirtualStringTree` it is executed. I recompiled the component with the `ShowMessage()` as you suggested. For C++ Builder program which also uses `TVirtualStringTree` it is not shown.

I used `ShowMessage()` and the parts between `begin` and `end;` are executed immediately on program startup, and the `TVTDefaultAccessibleProvider.CreateIAccessible` and `TVTMultiColumnAccessibleItemProvider.CreateIAccessible` are called the moment when screenreader requests it. After that, screenreader reads the listview fields properly.

For C++ Builder, none of the above is called, no `ShowMessage()` never appears. And as a result, screenreader doesn't read the listview control properly. So for C++ Builder it is definitely not initialized.

I am using NVDA for testing (https://www.nvaccess.org/) as it has a nice "speech viewer" window (right click NVDA tray icon and under the Tools menu there is `Speech Viewer` which logs exactly what the screenreader fetches to be read in a visual form (text), so you don't have to listen to the voice. Actual person with accessibility requirement would actually listen to the voice instead, but the speech viewer is good for debugging purposes.

**Comment by joachimmarder** [comment]
In case the unit `VirtualTrees.Accessibility` this result is unexpected. We have seen some similar issues in the past, like #642, but without a solution. To me this seems to be a bug in C++ Builder and I recommend contacting the Embarcadero support. For my understanding the initialization section of Delphi units should always be executed.

**Comment by joachimmarder** [comment]
I extracted the code in the Initialization section into a new class procedure `TVirtualTreeAccessibility.RegisterDefaultAccessibleProviders()` that you should be able to call in your code as a workaround. Does this work?

**Comment by 7even11** [comment]
Yes, calling the `TVirtualTreeAccessibility::RegisterDefaultAccessibleProviders();` somewhere in the program to initialize Accessibility in C++ Builder does the trick now. In fact, I find this even more convenient as then I don't have to initialize accessibility if for any reason I don't want to. The `initialization` part is definitely not called by itself but calling it manually does the trick.

There is a related issue described here as to why the `initialization` is not called due to optimization:
https://stackoverflow.com/questions/33738153/the-initialization-part-is-not-called

Based on the above, I tried adding in cpp file:
`#pragma link "VirtualTrees.Accessibility"`

and that worked too, without the above call to `TVirtualTreeAccessibility::RegisterDefaultAccessibleProviders();`, so that's another solution there.

It might be possible to overcome this by using {$HPPEMIT} to generate some extra C++ Builder code maybe something like:
`{$HPPEMIT '#pragma link "VirtualTrees.Accessibility"'}`

But for me personally the above 2 are more than enough as a solution.

While testing all of this I found some different problem which affects Delphi too, but that is not really part of this issue (I might create a new one), as for me at least, this issue can be closed.
